### PR TITLE
Environment to enable out of proc attestation for SGX

### PR DESCRIPTION
**Pull request by CyanDevs**
This sets the environment for enabling out-of-proc attestation by default by ensuring availability of the AESM service and by setting SGX_AESM_ADDR=1. In-proc attestation can used by unsetting SGX_AESM_ADDR.

Notes:
- This PR includes upstream changes from https://github.com/openenclave/openenclave/pull/3920 which should be merged first.
- libsgx-quote-ex is installed as part of Ansible changes in another PR by Ryan: https://github.com/openenclave/openenclave/pull/3882
- Docker run commands needs /var/run/aesmd/aesm.socket to be made available to the container:
  For example: `--volume /var/run/aesmd/aesm.socket:/var/run/aesmd/aesm.socket`
- This also fixes a RHEL 8 build issue where an out-of-date Docker package could no longer be installed.

Signed-off-by: Chris Yan chrisyan@microsoft.com

### CODE REVIEW

File: .jenkins/infrastructure/dockerfiles/linux/Dockerfile.minimal
Comment (anakrish) [inline .jenkins/infrastructure/dockerfiles/linux/Dockerfile.minimal:19]: Is there any significant benefit to using ansible within docker instead of just using apt-get?

Comment (CyanDevs) [reply .jenkins/infrastructure/dockerfiles/linux/Dockerfile.minimal:19]: It's best practice to leave container builds as simple as possible (e.g. without Ansible) but factoring out Ansible wasn't a consideration in the scope of this change. I do want to further optimize these docker builds, but it will have to be done separate to this change.

**Review by rs-- — APPROVED** [review]

**Review by radhikaj — APPROVED** [review]

**Comment by radhikaj** [comment]
Thanks @CyanDevs!

**Comment by CyanDevs** [comment]
Latest update was to remove an extra Ansible task from ci-setup.yml that ended up being unnecessary.

File: .jenkins/infrastructure/dockerfiles/linux/Dockerfile.full
Comment (yentsanglee) [inline .jenkins/infrastructure/dockerfiles/linux/Dockerfile.full:96]: These lines make developers environment safe from the SGX in-kernel transition. However, they also make the developers unaware of the fact that their application users have to do the same.

File: scripts/ansible/roles/linux/intel/tasks/sgx-driver.yml
Comment (yentsanglee) [inline scripts/ansible/roles/linux/intel/tasks/sgx-driver.yml:65]: These lines make developers environment safe from the SGX in-kernel transition. However, they also make the developers **unaware** of the fact that their application users have to do the same.

I suggest we remove these settings.

@bodzhang any comment?

**Comment by CyanDevs** [comment]
Latest changes:
* Rebased onto latest commit on master to pick up a CI fixes
* Included entrypoint overrides for Agnostic Linux builds and Azure Windows builds.

Tests done:
* [Building Docker images](https://oe-jenkins-tf.westeurope.cloudapp.azure.com/blue/organizations/jenkins/CI-CD_Infrastructure%2FOpenEnclave-Build-Docker-Images/detail/OpenEnclave-Build-Docker-Images/590/pipeline/
)
* [Building Azure VM Images](https://oe-jenkins-tf.westeurope.cloudapp.azure.com/blue/organizations/jenkins/CI-CD_Infrastructure%2FOpenEnclave-Build-Azure-Managed-Images/detail/OpenEnclave-Build-Azure-Managed-Images/733/pipeline/190/)
* [Azure-Linux using the new images](https://oe-jenkins-dev.westeurope.cloudapp.azure.com/blue/organizations/jenkins/pipelines%2FAzure-Linux/detail/Azure-Linux/2686/pipeline/) -- -- Note that snmalloc failures are a known existing issue and is being investigated separately
* [Azure-Windows](https://oe-jenkins-dev.westeurope.cloudapp.azure.com/blue/organizations/jenkins/pipelines%2FAzure-Windows/detail/Azure-Windows/2657/pipeline/398/)
* [Agnostic-Linux](https://oe-jenkins-dev.westeurope.cloudapp.azure.com/blue/organizations/jenkins/pipelines%2FAgnostic-Linux/detail/Agnostic-Linux/2599/pipeline/)

File: scripts/ansible/roles/linux/intel/tasks/sgx-driver.yml
Comment (CyanDevs) [reply scripts/ansible/roles/linux/intel/tasks/sgx-driver.yml:65]: It was actually Bo that suggested we put the environment variable into /etc/environment :)

**Comment by bodzhang** [comment]
This patch installs Intel's SGX DCAP driver 1.41, which mimic Linux kernel 5.11 native SGX support behavior/restriction. Is SGX DCAP driver 1.41 applied to all test flows, or it's only relevant to verify OE and Intel's latest SGX SW stack interaction with Linux kernel 5.11 native SGX support behavior?

**Comment by CyanDevs** [comment]
> This patch installs Intel's SGX DCAP driver 1.41, which mimic Linux kernel 5.11 native SGX support behavior/restriction. Is SGX DCAP driver 1.41 applied to all test flows, or it's only relevant to verify OE and Intel's latest SGX SW stack interaction with Linux kernel 5.11 native SGX support behavior?

The drivers have been updated on RHEL and Ubuntu VMs that are built for CI, so all tests that run on those will be using 1.41. E2E tests on CI have not had a problem with this, but do you feel that this is a concern?

**Comment by CyanDevs** [comment]
Last changes are to revert back to Intel SGX DCAP driver v1.35 to ensure we cover the most popular ACC configuration in our CI tests. We will look at 1.41 support at a later date.

File: scripts/ansible/roles/linux/intel/tasks/sgx-driver.yml
Comment (bodzhang) [reply scripts/ansible/roles/linux/intel/tasks/sgx-driver.yml:65]: We want to encourage users to move to out-of-proc attestation. I think configure it as default is OK.

**Review by bodzhang — APPROVED** [review]

**Comment by CyanDevs** [comment]
Images are built and pushed up to CI so this is being merged now.

**Comment by CyanDevs** [comment]
Will wait until #3882 is ready and we will merge it together.

**Comment by CyanDevs** [comment]
bors try

**Comment by CyanDevs** [comment]
bors try

**Comment by CyanDevs** [comment]
bors try

**Pull request merged**
